File: generator/src/mcexam/select.py
"""Deterministic variant construction from a seed. Mirror of grader/lib/select.dart.

RNG stream order (the contract — see README 'Determinism & reproducibility'):
all section selections first, in [easy, medium, hard] order, then all option
scrambles in sheet order. Do not reorder or lazily evaluate.
"""

from dataclasses import dataclass

from .model import SECTION_KEYS
from .rng import SplitMix64


@dataclass(frozen=True)
class SheetQuestion:
    section: str  # "easy" | "medium" | "hard"
    index_in_section: int  # 0-based index of the original question within its section
    global_index: int  # 0-based index in source order (easy ++ medium ++ hard)
    option_perm: tuple[int, ...]  # sheet position p shows original option option_perm[p]


@dataclass(frozen=True)
class VariantPlan:
    seed: int
    selections: dict[str, tuple[int, ...]]  # per section, in sheet order
    sheet: tuple[SheetQuestion, ...]


def build_variant(
    seed: int,
    section_sizes: dict[str, int],
    counts: dict[str, int],
    options_per_question: int,
) -> VariantPlan:
    for key in SECTION_KEYS:
        if not 0 <= counts[key] <= section_sizes[key]:
            raise ValueError(
                f"requested {counts[key]} '{key}' questions but the section has"
                f" only {section_sizes[key]}"
            )

    rng = SplitMix64(seed)

    selections: dict[str, tuple[int, ...]] = {}
    for key in SECTION_KEYS:  # all selections first ...
        indices = list(range(section_sizes[key]))
        rng.shuffle(indices)
        selections[key] = tuple(indices[: counts[key]])

    offsets: dict[str, int] = {}
    running = 0
    for key in SECTION_KEYS:
        offsets[key] = running
        running += section_sizes[key]

    sheet: list[SheetQuestion] = []
    for key in SECTION_KEYS:  # ... then option scrambles, in sheet order
        for index in selections[key]:
            perm = list(range(options_per_question))
            rng.shuffle(perm)
            sheet.append(
                SheetQuestion(
                    section=key,
                    index_in_section=index,
                    global_index=offsets[key] + index,
                    option_perm=tuple(perm),
                )
            )
    return VariantPlan(seed=seed, selections=selections, sheet=tuple(sheet))


def variant_seeds(base_seed: int, count: int) -> list[int]:
    """Per-variant seeds derived from the base seed (generator-internal: the
    actual per-variant seed always travels in the QR, so the grader never
    needs this derivation)."""
    rng = SplitMix64(base_seed)
    return [rng.next_uint64() for _ in range(count)]

File: grader/lib/select.dart
/// Deterministic variant construction from a seed. Mirror of
/// generator/src/mcexam/select.py.
///
/// RNG stream order (the contract — README 'Determinism & reproducibility'):
/// all section selections first, in [easy, medium, hard] order, then all
/// option scrambles in sheet order. Do not reorder or lazily evaluate.
library;

import 'rng.dart';

const List<String> sectionKeys = ['easy', 'medium', 'hard'];

class SheetQuestion {
  const SheetQuestion({
    required this.section,
    required this.indexInSection,
    required this.globalIndex,
    required this.optionPerm,
  });

  final String section;
  final int indexInSection;
  final int globalIndex;

  /// Sheet position p shows original option optionPerm[p].
  final List<int> optionPerm;
}

class VariantPlan {
  const VariantPlan({
    required this.seed,
    required this.selections,
    required this.sheet,
  });

  final BigInt seed;
  final Map<String, List<int>> selections;
  final List<SheetQuestion> sheet;
}

VariantPlan buildVariant({
  required BigInt seed,
  required Map<String, int> sectionSizes,
  required Map<String, int> counts,
  required int optionsPerQuestion,
}) {
  for (final key in sectionKeys) {
    final count = counts[key]!;
    final size = sectionSizes[key]!;
    if (count < 0 || count > size) {
      throw ArgumentError(
        "requested $count '$key' questions but the section has only $size",
      );
    }
  }

  final rng = SplitMix64(seed);

  final selections = <String, List<int>>{};
  for (final key in sectionKeys) {
    // All selections first ...
    final indices = List<int>.generate(sectionSizes[key]!, (i) => i);
    rng.shuffle(indices);
    selections[key] = indices.sublist(0, counts[key]!);
  }

  final offsets = <String, int>{};
  var running = 0;
  for (final key in sectionKeys) {
    offsets[key] = running;
    running += sectionSizes[key]!;
  }

  final sheet = <SheetQuestion>[];
  for (final key in sectionKeys) {
    // ... then option scrambles, in sheet order.
    for (final index in selections[key]!) {
      final perm = List<int>.generate(optionsPerQuestion, (i) => i);
      rng.shuffle(perm);
      sheet.add(
        SheetQuestion(
          section: key,
          indexInSection: index,
          globalIndex: offsets[key]! + index,
          optionPerm: perm,
        ),
      );
    }
  }
  return VariantPlan(seed: seed, selections: selections, sheet: sheet);
}
